apple-codesign/src/windows.rs

```diff
@@ -0,0 +1,736 @@
+            warn!(
+                "attempting to create signature using Windows store item: {}",
+                cn
+            );
+        }
+
+        // We need to determine the hash algorithm.
+        let hash_algorithm = match algorithm {
+            SignatureAlgorithm::RsaSha1 => BCRYPT_SHA1_ALGORITHM,
+            SignatureAlgorithm::RsaSha256 => BCRYPT_SHA256_ALGORITHM,
+            SignatureAlgorithm::RsaSha384 => BCRYPT_SHA384_ALGORITHM,
+            SignatureAlgorithm::RsaSha512 => BCRYPT_SHA512_ALGORITHM,
+            SignatureAlgorithm::EcdsaSha256 => BCRYPT_SHA256_ALGORITHM,
+            SignatureAlgorithm::EcdsaSha384 => BCRYPT_SHA384_ALGORITHM,
+            SignatureAlgorithm::Ed25519 => {
+                return Err(signature::Error::from_source("ed25519 not supported on windows"));
+            }
+            SignatureAlgorithm::NoSignature(_) => {
+                return Err(signature::Error::from_source("digest only signature"));
+            }
+        };
+
+        // We need to set the padding for NcryptSignHash. 
+        // Note that ECDSA signatures do not need
+        // padding, so it is fine to set this to null.
+        let (padding_info, flags) = match algorithm {
+            SignatureAlgorithm::RsaSha1 => {
+                (&mut BCRYPT_PKCS1_PADDING_INFO {
+                    pszAlgId: BCRYPT_SHA1_ALGORITHM as *mut u16,
+                } as *mut BCRYPT_PKCS1_PADDING_INFO, NCRYPT_PAD_PKCS1_FLAG)
+            }
+            SignatureAlgorithm::RsaSha256 => {
+                (&mut BCRYPT_PKCS1_PADDING_INFO {
+                    pszAlgId: BCRYPT_SHA256_ALGORITHM as *mut u16,
+                } as *mut BCRYPT_PKCS1_PADDING_INFO, NCRYPT_PAD_PKCS1_FLAG)
+            }
+            SignatureAlgorithm::RsaSha384 => {
+                (&mut BCRYPT_PKCS1_PADDING_INFO {
+                    pszAlgId: BCRYPT_SHA384_ALGORITHM as *mut u16,
+                } as *mut BCRYPT_PKCS1_PADDING_INFO, NCRYPT_PAD_PKCS1_FLAG)
+            }
+            SignatureAlgorithm::RsaSha512 => {
+                (&mut BCRYPT_PKCS1_PADDING_INFO {
+                    pszAlgId: BCRYPT_SHA512_ALGORITHM as *mut u16,
+                } as *mut BCRYPT_PKCS1_PADDING_INFO, NCRYPT_PAD_PKCS1_FLAG)
+            }
+            SignatureAlgorithm::EcdsaSha256 => {
+                (ptr::null_mut() as *mut BCRYPT_PKCS1_PADDING_INFO, 0)
+            }
+            SignatureAlgorithm::EcdsaSha384 => {
+                (ptr::null_mut() as *mut BCRYPT_PKCS1_PADDING_INFO, 0)
+            }
+            SignatureAlgorithm::Ed25519 => {
+                return Err(signature::Error::from_source("ed25519 not supported on windows"));
+            }
+            SignatureAlgorithm::NoSignature(_) => {
+                return Err(signature::Error::from_source("digest only signature"));
+            }
+        };
+
+        // We create a digest of the message using the BCrypt API.
+        let hash = try_hash(hash_algorithm, message)?;
```

Comment:
There is already code in the x509-certificate crate for computing digests from a `DigestAlgorithm` instance. So I'd like to think we can avoid calling into the win32 API for computing the digest and doing it all in Rust. See code in `yubikey.rs` for an example of how this is done.

The tricky part here is you need to look at the RSA key length and call `rsa_pkcs1_encode()` instead of `digest_data()` on the `DigestAlgorithm` instance.

(I should teach the `x509-certificate` crate to expose a one-stop digesting API that does all this for you to make things simpler.)